### PR TITLE
Document string builtin functions and string format

### PR DESCRIPTION
**Pull request by allredj**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d3a4d9dfe63f808bbe4b39097959ffc9e4c1836e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115994870

File: src/solvers/strings/string_builtin_function.h [hunk 1]
```diff
@@ -15,7 +15,59 @@ class string_constraint_generatort;
...
+/// Only the content of the strings are managed in that way. The constraints
+/// for the lengths are always added.
+///
+/// Concretely, if a builtin function supports **constraint generation**, it
+/// means that the result of the operation can be tested. For instance in this
+/// Java example:
+/// ```java
+/// String s = "foo" + "bar";
+/// if (s.equals("foo bar") {
+///   assert false;
```
Comment (peterschrammel) [inline src/solvers/strings/string_builtin_function.h:35]: that's not reachable (missing space)

File: src/solvers/strings/string_builtin_function.h [hunk 2]
```diff
@@ -15,7 +15,59 @@ class string_constraint_generatort;
...
+/// In this case, if concatenation supports evaluation, then the constraints for
+/// the concatenation operations are not added, but the concatenation is
+/// evaluated after solving, from the model, to render the concatenated string
+/// which is needed for generating the trace.
+///
+/// If the builtin function supports **constraint generation for the string
+/// length**, it means that the length of the string can be tested, as in:
+/// ```java
+/// String s = "foo" + "bar";
+/// if (s.length() == 7) {
```
Comment (peterschrammel) [inline src/solvers/strings/string_builtin_function.h]: see above

File: src/solvers/strings/string_builtin_function.h [hunk 3]
```diff
@@ -15,7 +15,59 @@ class string_constraint_generatort;
...
+/// The \ref string_dependenciest framework handles string operations by either:
+/// 1. generating constraints for the solver, which is done by the
+///    `constraints()` function,
+/// 2. evaluating constant operations from the solver model, which is done by
+///    the `eval()` function.
+///
+/// Only the content of the strings are managed in that way. The constraints
+/// for the lengths are always added.
+///
+/// Concretely, if a builtin function supports **constraint generation**, it
```
Comment (peterschrammel) [inline src/solvers/strings/string_builtin_function.h]: built-in (more of this)

File: src/solvers/strings/string_format_builtin_function.cpp [hunk 4]
```diff
@@ -1,15 +1,13 @@
 /*******************************************************************\
 
-Module: Builtin functions for String.format
+Module: Builtin function for String.format
```
Comment (peterschrammel) [inline src/solvers/strings/string_format_builtin_function.cpp]: Built-in

**Review by peterschrammel — APPROVED** [review]

File: src/solvers/strings/string_builtin_function.h [hunk 5]
```diff
@@ -15,7 +15,59 @@ class string_constraint_generatort;
...
+/// Base class for string functions that are built in the solver.
+///
+/// The \ref string_dependenciest framework handles string operations by either:
+/// 1. generating constraints for the solver, which is done by the
+///    `constraints()` function,
+/// 2. evaluating constant operations from the solver model, which is done by
+///    the `eval()` function.
+///
+/// Only the content of the strings are managed in that way. The constraints
+/// for the lengths are always added.
```
Comment (romainbrenguier) [inline src/solvers/strings/string_builtin_function.h]: explicitly mention the `length_constraint()` function

File: src/solvers/strings/string_builtin_function.h [hunk 6]
```diff
@@ -15,7 +15,59 @@ class string_constraint_generatort;
...
+/// means that the result of the operation can be tested. For instance in this
+/// Java example:
+/// ```java
+/// String s = "foo" + "bar";
+/// if (s.equals("foo bar") {
+///   assert false;
+/// }
+/// ```
+/// if the builtin function for concatenation supports constraint generation,
+/// then the assertion is reachable.
```
Comment (romainbrenguier) [inline src/solvers/strings/string_builtin_function.h]: I'm not sure I understand. If the constraints are not added, then more behaviors are allowed so the assertion should be reachable as well.

**Review by romainbrenguier — COMMENTED** [review]
The table is very good but it does not mention precision, width, tags, indexes

File: src/solvers/strings/string_builtin_function.h (hunk 1 above)
Comment (allredj) [reply src/solvers/strings/string_builtin_function.h:35]: fixed

File: src/solvers/strings/string_builtin_function.h (hunk 2 above)
Comment (allredj) [reply src/solvers/strings/string_builtin_function.h]: fixed

File: src/solvers/strings/string_builtin_function.h (hunk 3 above)
Comment (allredj) [reply src/solvers/strings/string_builtin_function.h]: fixed everywhere

File: src/solvers/strings/string_format_builtin_function.cpp (hunk 4 above)
Comment (allredj) [reply src/solvers/strings/string_format_builtin_function.cpp]: fixed

File: src/solvers/strings/string_builtin_function.h (hunk 5 above)
Comment (allredj) [reply src/solvers/strings/string_builtin_function.h]: done

File: src/solvers/strings/string_builtin_function.h (hunk 6 above)
Comment (allredj) [reply src/solvers/strings/string_builtin_function.h]: added an `else` to the example

**Comment by allredj** [comment]
@romainbrenguier, line about precision, width, tags, indexes components has been added.

**Review by romainbrenguier — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 67c7e7d39fd98eb8f87c0de7377ddebe11bd657d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117127437
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Pull request merged**